obo_parser.py

```diff
@@ -0,0 +1,388 @@
+RENAME_COLUMNS = {
+    'is_a': 'parent_ids',
+    'def': 'definition',
+}
+
+
+def convert_obo_to_tsv(input_path, output_path="-", root_id=None, add_category_column=False):
+    """Main entry point for parsing an .obo file and converting it to a .tsv table.
+
+    Args:
+        input_path (str): .obo file url or local file path
+        output_path (str): path where to write the .tsv file. Defaults to "-" which is standard out.
+        root_id (str): If specified, ignore ontology terms that are not either descendants of the
+            given id or have this id themselves. For example, 'HP:0000118'.
+        add_category_column (bool): Whether to add a 'category' column to the output .tsv file
+            which lists each term's top-level category. A top-level category is a term that's a
+            direct child of the ontology's root term.
+    """
+
+    if output_path is None:
+        output_path = os.path.basename(input_path).replace(".obo", "") + ".tsv"
+
+    # read in data
+    logger.info("Parsing %s", input_path)
+    with _open_input_stream(input_path) as input_stream:
+        obo_records_dict = parse_obo_format(input_stream)
+
+    # find root term
+    if root_id is None:
+        root_id = _compute_root_id(obo_records_dict)
+
+    _confirm_id_is_valid(root_id, obo_records_dict, label="root_id")
+
+    # add 'category' columns to records
+    if add_category_column:
+        compute_category_column(obo_records_dict, root_id=root_id)
+
+    # print stats and output .tsv
+    print_stats(obo_records_dict, input_path)
+
+    if output_path == "-":
+        write_tsv(obo_records_dict, output_stream=sys.stdout, root_id=root_id)
+    else:
+        with open(output_path, "w") as output_stream:
+            write_tsv(obo_records_dict, output_stream, root_id=root_id)
+
+    logger.info("Done")
+
+
+def parse_obo_format(lines):
+    """Parses .obo-formatted text.
+
+    Args:
+        lines (iter): Iterator over lines of text in .obo format.
+    Returns:
+        dict: .obo records, keyed by term id. Each record is a dictionary where the keys are tags
+            such as "id", "name", "is_a", and values are strings (for tags that can only occur once
+             - such as "id"), or lists (for tags that can appear multiple times per stanza - such as
+             "xref")
+    """
+
+    obo_records_dict = collections.OrderedDict()
+    current_stanza_type = None
+    current_record = None
+    if logger.isEnabledFor(logging.INFO):
+        lines = tqdm.tqdm(lines, unit=" lines")
+
+    for line in lines:
+        if line.startswith("["):
+            current_stanza_type = line.strip("[]\n")
+            continue
+
+        # skip header lines and stanzas that aren't "Terms"
+        if current_stanza_type != "Term":
+            continue
+
+        # remove new-line character and any comments
+        line = line.rstrip('\n').split("!")[0]
```

Comment:
Is whitespace generally important on the edges? Generally, `line.strip()` will catch any other weird whitespace issues.